resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -142,16 +149,30 @@ public void operationComplete(Future<AddressedEnvelope<DnsResponse, InetSocketAd
         });
     }
 
-    void onResponse(final DnsQuestion question, AddressedEnvelope<DnsResponse, InetSocketAddress> response) {
-        final DnsRecordType type = question.type();
+    void onResponse(final DnsQuestion question, AddressedEnvelope<DnsResponse, InetSocketAddress> envelope) {
         try {
-            if (type == DnsRecordType.A || type == DnsRecordType.AAAA) {
-                onResponseAorAAAA(type, question, response);
-            } else if (type == DnsRecordType.CNAME) {
-                onResponseCNAME(question, response);
+            final DnsResponse res = envelope.content();
+            final DnsResponseCode code = res.code();
+            if (code == DnsResponseCode.NOERROR) {
+                final DnsRecordType type = question.type();
+                if (type == DnsRecordType.A || type == DnsRecordType.AAAA) {
+                    onResponseAorAAAA(type, question, envelope);
+                } else if (type == DnsRecordType.CNAME) {
+                    onResponseCNAME(question, envelope);
+                }
+                return;
+            }
+
+            addTrace(envelope.sender(),
```

Comment:
should we add something like "isTraceEnabled()" so we only need to do this if tracing is requested ?